### PR TITLE
set reentrancy block when call trade during claim platformFee and rebate

### PR DESCRIPTION
**Pull request by ducquangkstn**
Case:
- The platform has the fee of 5 wei
- call **claimPlatformFee** -> receive 4 wei
- trigger callback function -> Earn some more platform fees
- PlatformFeePaid emitted.
-> platform fee is not reset to 1

In this PR:
- add reentrancy guard to claimPlatformFee, handleFees, claimReserveRebate
- add claimPlatformFee to IKyberFeeHandler

### CODE REVIEW

File: contracts/sol6/Dao/KyberFeeHandler.sol [hunk 1]
```diff
@@ -261,7 +262,7 @@ contract KyberFeeHandler is IKyberFeeHandler, Utils5, DaoOperator {
     /// @dev claim accumulated fee per platform wallet. Called by any address
     /// @param platformWallet the wallet to claim fee for. Total accumulated fee sent to this wallet.
     /// @return feeWei amount of fee claimed
-    function claimPlatformFee(address platformWallet) external returns (uint256 feeWei) {
+    function claimPlatformFee(address platformWallet) external override nonReentrant returns (uint256 feeWei) {
```
Comment (ilanDoron) [inline contracts/sol6/Dao/KyberFeeHandler.sol:265]: notice to limit to 99 chars per line

prettier is off.
but will add it back soon.

File: contracts/sol6/mock/ReentrancyFeeClaimer.sol [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        if (count != 0) {
+            return;
+        }
+
+        if (!isReentrancy) {
+            return;
+        }
+
+        bytes memory hint;
+        kyberProxy.tradeWithHintAndFee(
```
Comment (ilanDoron) [inline contracts/sol6/mock/ReentrancyFeeClaimer.sol:54]: can use one of the simple swap API

File: contracts/sol6/mock/ReentrancyFeeClaimer.sol [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    function claimReserveRebate() external {
+        count = 0;
+        feeHandler.claimReserveRebate(address(this));
+    }
+
+    function setReentrancy(bool _isReentrancy) external {
+        isReentrancy = _isReentrancy;
+    }
+
+    receive() external payable {
```
Comment (ilanDoron) [inline contracts/sol6/mock/ReentrancyFeeClaimer.sol:44]: isn't receive payable by default?

File: test/sol6/kyberFeeHandler.js [hunk 4]
```diff
@@ -1562,6 +1566,107 @@ contract('KyberFeeHandler', function(accounts) {
...
+                    claimer.claimPlatformFee(),
+                    "platform fee transfer failed"
+                );
+                // if we not reentrancy, then still claim fee
+                await claimer.setReentrancy(false);
+                await claimer.claimPlatformFee();
+                await claimer.setReentrancy(true);
+            })
+
+            it("revert if call trade during claim platform fee", async() => {
```
Comment (ilanDoron) [inline test/sol6/kyberFeeHandler.js:1652]: should be during claim rebate

File: contracts/sol6/mock/ReentrancyFeeClaimer.sol [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        if (count != 0) {
+            return;
```
Comment (ilanDoron) [inline contracts/sol6/mock/ReentrancyFeeClaimer.sol:46]: why do we need this if?

if count > 0 it should fail when trade is called
no?

File: contracts/sol6/mock/ReentrancyFeeClaimer.sol (hunk 2 above)
Comment (ducquangkstn) [reply contracts/sol6/mock/ReentrancyFeeClaimer.sol:54]: using `tradeWithHintAndFee` will updates the platformFee to this address

File: contracts/sol6/mock/ReentrancyFeeClaimer.sol (hunk 3 above)
Comment (ducquangkstn) [reply contracts/sol6/mock/ReentrancyFeeClaimer.sol:33]: It must be payable, not payable by default
https://solidity.readthedocs.io/en/v0.6.0/contracts.html#receive-ether-function

File: contracts/sol6/Dao/KyberFeeHandler.sol (hunk 1 above)
Comment (ducquangkstn) [reply contracts/sol6/Dao/KyberFeeHandler.sol]: change as request

File: contracts/sol6/mock/ReentrancyFeeClaimer.sol (hunk 5 above)
Comment (ducquangkstn) [reply contracts/sol6/mock/ReentrancyFeeClaimer.sol]: removed it.

File: test/sol6/kyberFeeHandler.js (hunk 4 above)
Comment (ducquangkstn) [reply test/sol6/kyberFeeHandler.js]: change as request

File: contracts/sol6/mock/ReentrancyFeeClaimer.sol (hunk 3 above)
Comment (ilanDoron) [reply contracts/sol6/mock/ReentrancyFeeClaimer.sol:33]: I see thanks

File: test/sol6/kyberFeeHandler.js [hunk 6]
```diff
@@ -1562,6 +1566,107 @@ contract('KyberFeeHandler', function(accounts) {
                 Helper.assertEqual(0, await feeHandler.getLatestSanityRate());
             });
         });
+
+        describe("reentrancy lead to inaccurate event when claiming fee", async() =>{
```
Comment (ilanDoron) [inline test/sol6/kyberFeeHandler.js:1570]: I would write

test claim fee functions (claimRebate etc..) in fee handler are blocking re entrance calls.

File: test/sol6/kyberFeeHandler.js [hunk 7]
```diff
@@ -1562,6 +1566,107 @@ contract('KyberFeeHandler', function(accounts) {
...
+                await token.approve(networkProxy.address, srcQty, { from: taker});
+                await networkProxy.tradeWithHintAndFee(token.address, srcQty, ethAddress, taker,
+                    maxDestAmt, 0, claimer.address, new BN(100), '0x', {from: taker});
+                let platformFee = await feeHandler.feePerPlatformWallet(claimer.address);
+                Helper.assertGreater(platformFee, new BN(1), "platform fee should be enough to withdraw");
+                await expectRevert(
+                    feeHandler.claimPlatformFee(claimer.address),
+                    "platform fee transfer failed"
+                );
+                // if we not reentrancy, then still claim fee
```
Comment (ilanDoron) [inline test/sol6/kyberFeeHandler.js:1646]: maybe better:

when no re entrance, claim fee succeeds

File: contracts/sol6/Dao/KyberFeeHandler.sol [hunk 8]
```diff
@@ -238,7 +239,7 @@ contract KyberFeeHandler is IKyberFeeHandler, Utils5, DaoOperator {
     /// @dev claim reabate per reserve wallet. called by any address
     /// @param rebateWallet the wallet to claim rebates for. Total accumulated rebate sent to this wallet.
     /// @return amount of rebate claimed
-    function claimReserveRebate(address rebateWallet) external override returns (uint256) {
+    function claimReserveRebate(address rebateWallet) external override nonReentrant returns (uint256) {
```
Comment (ilanDoron) [inline contracts/sol6/Dao/KyberFeeHandler.sol:242]: also line too long.

File: test/sol6/kyberFeeHandler.js (hunk 6 above)
Comment (ducquangkstn) [reply test/sol6/kyberFeeHandler.js]: updated

File: test/sol6/kyberFeeHandler.js (hunk 7 above)
Comment (ducquangkstn) [reply test/sol6/kyberFeeHandler.js]: updated

File: contracts/sol6/Dao/KyberFeeHandler.sol (hunk 8 above)
Comment (ducquangkstn) [reply contracts/sol6/Dao/KyberFeeHandler.sol]: updated

**Pull request merged**